pkg/compiler/check.go

```diff
@@ -600,6 +608,10 @@ func (comp *compiler) checkTypeCtors(t *ast.Type, dir prog.Dir, isArg bool,
 		}
 		checked[key] = true
 		for _, fld := range s.Fields {
+			// If the field has direction, it overrides any upper levels
+			if hasDir, fldDir := comp.genFieldDir(fld); hasDir {
+				dir = fldDir
```

Comment:
This will affect all subsequent fields as well.
Need some way to test this, but I think it can only be reasonably tested on the prog package level by desalinizing some programs that will use directions, or something similar.

Reply:
[This](https://github.com/google/syzkaller/pull/2008/commits/82ffbda67776502531b39b560e33ac8956d90827#diff-47a4f07e1880cd23bc05dca73d329c1dL612-L615) should fix the bug.

Testing is on my agenda.